### ISSUE TITLE
Open issue: non-negative assumption should apply to both Count and Length

### ISSUE THREAD

**Issue by jcouv** [dotnet/csharplang#5137]
The current behavior is shown below. I think we should have an error for both lines.

```
interface IIndexable
{
    int this[int i] { get; }
}
interface ICountableViaCount
{
    int Count { get; }
}
interface ICountableViaLength
{
    int Length { get; }
}
class X
{
    public static void Test5<T>(T t) where T : IIndexable, ICountableViaLength, ICountableViaCount
    {
        _ = t is { Length: -1 }; // error CS8518: An expression of type 'T' can never match the provided pattern.
        _ = t is { Count: -1 }; // ok
    }
}
```

FYI @alrz @333fred 

Relates to proposal https://github.com/dotnet/csharplang/issues/3435 
Relates to test plan https://github.com/dotnet/roslyn/issues/51289

**Comment by JustNrik** [comment]
Something I've always been curious about, why are `Length` and `Count` declared as `int` when they are really never going to be negative? like, shouldn't they be `uint` instead? 

Also, why have two interfaces that are essentially the same when you could just have a single `ICountable` interface, or perhaps `ISizable` with `Size` property?

Should `IIndexable` really be readonly? couldn't we have `IReadOnlyIndexable` with `IIndexable : IReadOnlyIndexable` instead? Or are all of this interfaces just for list patterns features and not general abstraction?

**Comment by CyrusNajmabadi** [comment]
Unsigned integers were not cls compliant.

**Comment by theunrepentantgeek** [comment]
When the compiler encounters this line will it already be sure that it's working with a collection or sequence?

``` csharp
 _ = t is { Length: -1 }; // error CS8518: An expression of type 'T' can never match the provided pattern.
```

If the compiler can't be _completely_ sure that `t` is an **array** (or a similar container) then a warning here might be a problem due to differing semantics. Not every `Length` is strictly non-negative.

**Comment by CyrusNajmabadi** [comment]
Our position is that it's fine for the language to assume these types are well behaved for the purposes of pattern matching.

**Comment by FaustVX** [comment]
Why not put an attribute on the correct _length_ property (like `IndexableLength`)
This will allow to have any name for the property we want (`Count`, `Length`, `Size`, even other language `Longueur`)
Maybe with a parameter to allow negative value.

```csharp
class C {
  public object this[int i] { get; }
  [IndexableLength(allowNegative = false)]
  public int Size { get; }
}

_ = new C() is { Size: -1 }; // error
```

**Comment by CyrusNajmabadi** [comment]
> Why not put an attribute on the correct length property (like IndexableLength)

Because the APIs have already shipped, and we'd want patterns to work across the enormous ecosystem of well-behaving types already out there.

**Comment by 333fred** [comment]
> When the compiler encounters this line will it already be sure that it's working with a collection or sequence?

Yes, it will only make this assumption when the type is both Countable _and_ Indexable.

**Comment by alrz** [comment]
I suppose LongLength would be out of scope here because it doesn't have any special meaning.

**Comment by jcouv** [comment]
The former "If both are present, only one of them is assumed to be non-negative". If there is a Length, then we say/assume nothing about Count.
